esss_fix_format/code_formatter.py

```diff
@@ -0,0 +1,421 @@
+'''
+Module to handle code formatting.
+
+There are 3 modes of operation:

```

Comment:
Nice that there are these 3 modes!

I was a little bit worried about the complexity of the Mutex code that will have to be maintained now (possibility of race conditions, etc.), but I've tested it on all ALFA Sim code, and mode 1 is _really_ slow (2 minutes against 9 seconds).

Suggestion for a different mode of operation: could it run in a "serverless/daemonless" fashion, like, just calling `java` passing a list of all files to be formatted, and wait for `java` to complete?

Reply:
I think it'd still be annoying (because the first invocation will always pay around 6-7 seconds for the java warmup, whereas in the current mode it'll always be fast excluding the first call after you reboot the machine)... and those few seconds each time you're about to commit or do a fix-format will be very annoying IMHO.

Reply:
Hmm, yeah, JVM is sometimes really slow to startup. I did a quick search about other solutions for this, and noticed that there is this [Nailgun project](http://martiansoftware.com/nailgun/). It seems to implement a server also, but one that can be shared (with the bonus that we wouldn't have to maintain it, hehe).

Anyway, I didn't investigate any further, just sharing in case you'd like to give a try.

Reply:
Well, I took a brief look at it -- also found: https://github.com/ninjudd/drip but it seems that each of those comes with its own set of shortcomings (as well as new dependencies which may themselves be hard to manage).